---
title: "ComPlEx - Asp_NorWood - copy"
author: "Ellen Dimmen Chapple"
date: "2023-08-28"
output: html_document

editor_options: 
  chunk_output_type: console
---

```{r}
# Species name in Plaza
species1_name <- "Asp"

species2_name <- "Nor"

# Expression files
species1_expr_file <- "Data/DATA/AspWood_transcriptomics.txt"

species2_expr_file <- "Data/DATA/NorWood_transcriptomics.txt"
# Ortholog groups

ortholog_group_file <- "Data/DATA/Orthogroups.100323.tsv"

cat("ComPlEX:", species1_name, species2_name, "\n")
```

```{r}
# Read in ortholog groups
# =======================

ortholog_group_RData <- paste0("Data/DATA/orthologs-", species1_name, "-", species2_name, ".RData")

if (!file.exists(ortholog_group_RData)) {  # Asp Nor
  ortho <- read.delim2(ortholog_group_file, header = TRUE, sep = "\t") %>%
    dplyr::rename(Species1 = Poptre.SHORT.pep, Species2 = Picabi.SHORT.pep, OrthoGroup = Orthogroup) %>%
    select(Species1, Species2, OrthoGroup) %>%
    filter(Species1 != "", Species2 != "") %>%
    separate_rows(Species1, sep = ", ", convert = FALSE) %>%
    separate_rows(Species2, sep = ", ", convert = FALSE) %>%
    mutate(Species1 = gsub("\\.\\d\\.p\\d$", "", Species1)) %>%
    mutate(Species2 = gsub("\\.p\\d$", "", Species2)) %>%
    mutate(Species1 = gsub("Poptre_", "", Species1)) %>%
    group_by(Species1, Species2, OrthoGroup) %>%
    dplyr::slice(1)
  
    # Add annotations from arabidopsis
  symbols <- read.delim("Data/DATA/gene_aliases_20140331.txt", sep = "\t") %>%
    dplyr::rename(Arab = locus_name, Symbol = symbol, Name = full_name)
  
  annot <- read.delim2(ortholog_group_file, header = TRUE, sep = "\t") %>%
    dplyr::rename(Arab = Aratha.SHORT.pep, OrthoGroup = Orthogroup) %>%
    select(Arab, OrthoGroup) %>%
    filter(Arab != "") %>%
    separate_rows(Arab, sep = ", ", convert = FALSE) %>%
    mutate(Arab = gsub("\\.\\d.p\\d$", "", Arab)) %>%
    mutate(Arab = gsub("Aratha_", "", Arab)) %>%
    left_join(symbols, by = "Arab") %>%
    group_by(OrthoGroup) %>%
    summarise(Arab = paste0(unique(Arab), collapse = "; "), 
           Symbol = paste0(unique(Symbol), collapse = "; "), 
           Name = paste0(unique(Name), collapse = "; ")) %>%
    mutate(Symbol = gsub("NA", "", Symbol),
           Name = gsub("NA", "", Name))

  save(ortho, annot, file = ortholog_group_RData)
} else {
  load(file = ortholog_group_RData)
}

# Read in expression data
# =======================

species1_expr <- read.delim(species1_expr_file, sep = "\t", header = TRUE)

species2_expr <- read.delim(species2_expr_file, sep = "\t", header = TRUE)

# Filter
# ======

cat (length(unique(ortho$OrthoGroup)), " ortholog groups containing:\n",
     " ", length(unique(ortho$Species1)), " ", species1_name, " genes\n",
     " ", length(unique(ortho$Species2)), " ", species2_name, " genes\n\n",
     length(unique(species1_expr$Genes)), " expressed ", species1_name, " genes\n",
     length(unique(species2_expr$Genes)), " expressed ", species2_name, " genes\n",
     sep = "")

```




```{r}

ortho <- ortho %>%
  filter(Species1 %in% species1_expr$Genes & Species2 %in% species2_expr$Genes)

species1_expr <- species1_expr[species1_expr$Genes %in% ortho$Species1,]
species2_expr <- species2_expr[species2_expr$Genes %in% ortho$Species2,]

cat ("After filtering on expressed genes with ortholog:\n",
     " ", length(unique(ortho$OrthoGroup)), " ortholog groups containing: \n",
     "  ", length(unique(ortho$Species1)), " ", species1_name, " genes\n",
     "  ", length(unique(ortho$Species2)), " ", species2_name, " genes\n",
     sep = "")

```






























