---
title: "Comparative heatmap"
author: "Ellen Dimmen Chapple"
date: "2023-08-25"
output: 
  html_document:
    code_folding: hide
    toc: true
    number_sections: true
    toc_float: true
    theme: paper
editor_options:
  chunk_output_type: console
---


```{r Packages, message=FALSE, warning=FALSE}
library(dplyr)
library(tidyr)
library(ggplot2)
library(gplots)
library(beepr)
library(reshape2)

```


**Flow**


- Load expression datasets for Aspen (AspWood) and Norway Spruce (NorWood).

- Load comparison table which contains the various genes within each ortholog group for both species. The p-value for each    ortholog pair is calculated (both directions), and then the highest of these two is used for 'Max.p.val'.

- Extract the lowest 'Max.p.val.' from each ortholog group with the corresponding ortholog pair.

- Filter the expression data to only include genes that are from the lowest p-value ortholog pairs. The rows in the           expression data sets must also be ordered the same so that the gene in row 10 in species 1 will have its ortholog pair in   row 10 in species 2.
  
- Create a correlation matrix using the filtered and sorted/ordered expression data.

- Order the dendrograms to follow an order similar to the samples.

- Create heatmap.



# Loading the transcription data



```{r}
# AspWood
asp_expr_data <- read.delim ("Data/DATA/AspWood_transcriptomics.txt")


# NorWood
nor_expr_data <- read.delim("Data/DATA/NorWood_transcriptomics.txt")

cat("AspWood - # of genes: ", dim(asp_expr_data)[1],"\nNorWood - # of genes: ", dim(nor_expr_data)[1])
```




```{r loaded comp tbl}
load("~/NMBU/M60-BIAS/EvoTree/Data/DATA/comparison-Asp-Nor-pearsonMR0.03no-table.RData")
```


# Selecting ortholog pairs with lowest max p-value

- 'Max.p.val' needs to be numeric.

- Use distinct() to avoid the same gene appearing in the different OGs - this will otherwise cause the filtered expression    data sets to have different row lengths which is needed for making correlation matrix. 


```{r complete set p val}

asp_nor_p.val <- comparison_table%>%
  ungroup() %>% 
  select(Species1, Species2,OrthoGroup, Max.p.val) %>%
  mutate_at("Max.p.val", as.numeric) %>% 
  group_by(OrthoGroup) %>% 
  arrange(Max.p.val,.by_group = T) %>% 
  slice(1) %>% 
  ungroup() %>% 
  distinct(Species1, .keep_all = T) %>% 
  distinct( Species2, .keep_all = T)
# Check the number of unique groups in test set.

cat("Number of orthogroups after removing genes appearing in multiple groups: ", length(unique(asp_nor_p.val$OrthoGroup)) )

cat("Number of duplicated genes in Asp: ", sum(duplicated(asp_nor_p.val$Species1)), "\nNumber of duplicated genes in Nor: ",sum(duplicated(asp_nor_p.val$Species2)))


# # Create a vector with the unique Orthogroups
# og_2s <- unique(loop_2s$OrthoGroup)
# species <- c(1,2)
# 
# beep(2)
# 
# # For each unique group, select lowest p-value, add to new df.
# for(x in og_2s){
#   for(i in species){
#   lowest_p_val_2s <- loop_2s %>% 
#   slice(which.min(Max.p.val))
# }
# }
# # Check that the number of groups is the same as in test data set.
# length(unique(lowest_p_val_2s$OrthoGroup))
# beep(3)

```

# Filtering the 

```{r complete set filtering}
# ----------------- ASP ------------------
# Filter the genes in the expression tables to only include the genes in the 'lowest_p_value_s' df.
asp_expr_filt_comp <- asp_expr_data %>% 
  filter(Genes %in% asp_nor_p.val$Species1) %>% 
  select(contains("A1."))

sum(duplicated(asp_expr_filt_comp$Genes))

# New data frame without first row (genes)
asp_expr_sort_comp <- asp_expr_filt_comp[,-1]
# Add genes as row names 
rownames(asp_expr_sort_comp) <- asp_expr_filt_comp$Genes

# Order the rows
asp_expr_sort_comp <-asp_expr_sort_comp[order(match (rownames(asp_expr_sort_comp), asp_nor_p.val$Species1)), , drop = F]

sum(duplicated(asp_expr_sort_comp$Genes))

# ----------------- Nor ------------------ 


nor_expr_filt_comp <- nor_expr_data %>% 
  filter(Genes %in% asp_nor_p.val$Species2) %>% 
  select(contains("S1."))

sum(duplicated(nor_expr_filt_comp$Genes))

# New data frame without first row (genes)
nor_expr_sort_comp <- nor_expr_filt_comp[,-1]
# Add genes as row names 
rownames(nor_expr_sort_comp) <- nor_expr_filt_comp$Genes

# Order the rows
nor_expr_sort_comp <-nor_expr_sort_comp[order(match (rownames(nor_expr_sort_comp), asp_nor_p.val$Species2)), , drop = F]

sum(duplicated(nor_expr_filt_comp$Genes))

```



```{r complete set correlation and heatmap}


corr_matrix <- round(cor(asp_expr_sort_comp, nor_expr_sort_comp, method = "pearson"), 4)

# Need to order the samples. We know that the clusters more or less fall into a sequential order.

#Aspen
dist_1 <- dist(t(asp_expr_sort_comp), method = "euclidean")
clust_asp <- hclust(dist_1, method = "ward.D2", members = NULL)
dend_sam_1 <- as.dendrogram(clust_asp)
asp_ordered <- reorder(dend_sam_1, 1:ncol(asp_expr_sort_comp))


#Norway Spurce
dist_2 <- dist(t(nor_expr_sort_comp), method = "euclidean")
clust_nor <- hclust(dist_2, method = "ward.D2", members = NULL)
dend_sam_2 <- as.dendrogram(clust_nor)
nor_ordered <- reorder(dend_sam_2, 1:ncol(nor_expr_sort_comp))

# corr_matrix : 105x39


#Heatmap

heat <- heatmap.2(corr_matrix, 
                  trace = "none",
                  col = colorRampPalette(c("blue","white","red")),
                  Rowv = FALSE,
                  Colv = FALSE,
                  main = "Aspen - Norway Spruce") 

```








