---
title: "Comparative heatmap"
author: "Ellen Dimmen Chapple"
date: "2023-08-25"
output: 
  html_document:
    code_folding: hide
    toc: true
    number_sections: true
    toc_float: true
    theme: paper
editor_options:
  chunk_output_type: console
---


```{r Packages, message=FALSE, warning=FALSE}
library(dplyr)
library(tidyr)
library(ggplot2)
library(ComplexHeatmap)
library(tibble)
library(reshape2)
library(cowplot)
library(grid)

```


**Flow**


- Load expression datasets for: Aspen (AspWood) Norway Spruce (NorWood) and Scots Pine (ScotsWood)

- Load comparison table which contains the various genes within each ortholog group for both species. The p-value for each    ortholog pair is calculated (both directions), and then the highest of these two is used for 'Max.p.val'.

- Extract the lowest 'Max.p.val.' from each ortholog group with the corresponding ortholog pair.

- Filter the expression data to only include genes that are from the lowest p-value ortholog pairs. The rows in the           expression data sets must also be ordered the same so that the gene in row 10 in species 1 will have its ortholog pair in   row 10 in species 2.
  
- Create a correlation matrix using the filtered and sorted/ordered expression data.

- Create heatmap.



# Loading the transcription data



```{r}
# AspWood
asp_expr_data <- read.delim ("Data/DATA/AspWood_transcriptomics.txt")


# NorWood
nor_expr_data <- read.delim("Data/DATA/NorWood_transcriptomics.txt")

# ScotsWood (Pinus Sylvestris, Scots Pine, Furu)
scots_expr_data <- read.delim("Data/DATA/ScotsWood_transcriptomics.txt")

cat("AspWood - # of genes: ", dim(asp_expr_data)[1],"\nNorWood - # of genes: ", dim(nor_expr_data)[1], "\nScotsWood - # of genes: ", dim(scots_expr_data)[1])
```



# Selecting ortholog pairs with lowest max p-value

- 'Max.p.val' needs to be numeric.

- Use distinct() to avoid the same gene appearing in the different OGs - this will otherwise cause the filtered expression data sets to have different row lengths which is needed for making correlation matrix. However, must be used after picking out the lowest p-values!

NOTES FOR CODE BELOW: Need more compact code that can be used for all species combinations.


```{r ASPNOR}



load("Data/DATA/comparison-Asp-Nor-pearsonMR0.03no-table.RData")
asp_nor_p.val <- comparison_table%>%
  # ungroup() %>% 
  select(Species1, Species2,OrthoGroup, Max.p.val) %>%
  mutate_at("Max.p.val", as.numeric) %>% 
  group_by(OrthoGroup) %>% 
  arrange(Max.p.val,.by_group = T) %>% 
  slice(1) %>% 
  ungroup() %>% 
  distinct(Species1, .keep_all = T) %>% 
  distinct( Species2, .keep_all = T)


# Check the number of unique groups in test set.
cat("Number of orthogroups after removing genes appearing in multiple groups: ", length(unique(asp_nor_p.val$OrthoGroup)) )

cat("Number of duplicated genes in Asp: ", sum(duplicated(asp_nor_p.val$Species1)), "\nNumber of duplicated genes in Nor: ",sum(duplicated(asp_nor_p.val$Species2)))

# New filtering and sorting - ASP
asp_pval_copy <- asp_nor_p.val

asp_expr_fs <- asp_pval_copy %>% 
  left_join(asp_expr_data, join_by(Species1 == Genes)) %>% 
  select(-c(2:4)) %>% 
  as_tibble() %>% 
  column_to_rownames(var = "Species1") %>% 
  select(contains("1."))
  

# New filtering and sorting - NOR

nor_pval_copy <- asp_nor_p.val
nor_expr_fs <- nor_pval_copy %>% 
  left_join(nor_expr_data, join_by(Species2 == Genes)) %>% 
  select(-c("Species1", "Max.p.val", "OrthoGroup")) %>% 
  as_tibble() %>% 
  column_to_rownames(var = "Species2") %>% 
  select(contains("1."))
# Correlation matrix and heatmap

corr_asp_nor <- round(cor( asp_expr_fs,nor_expr_fs, method = "pearson"), 4)
heat_asp_nor <-  Heatmap(corr_asp_nor, column_title = "Asp-Nor",
    column_title_gp = gpar(fontsize = 15),
    row_names_gp = gpar(fontsize = 8),
    column_names_gp = gpar(fontsize = 8),
    width = ncol(corr_asp_nor)*unit(5, "mm"), 
    height = nrow(corr_asp_nor)*unit(5, "mm"),
    row_order = 1:ncol(asp_expr_fs),
    column_order = 1:ncol(nor_expr_fs),
    name = "corr" ) 

# https://jokergoo.github.io/ComplexHeatmap-reference/book/

```



```{r ASPSCOTS}


#------------------ASP-SCOTS------------------
load("Data/DATA/comparison-Asp-Scots-pearsonMR0.03no-table.RData")


asp_scots_p.val <- comparison_table%>%
  # ungroup() %>% 
  select(Species1, Species2,OrthoGroup, Max.p.val) %>%
  mutate_at("Max.p.val", as.numeric) %>% 
  group_by(OrthoGroup) %>% 
  arrange(Max.p.val,.by_group = T) %>% 
  slice(1) %>% 
  ungroup() %>% 
  distinct(Species1, .keep_all = T) %>% 
  distinct( Species2, .keep_all = T)
# Check the number of unique groups in test set.

cat("Number of orthogroups after removing genes appearing in multiple groups: ", length(unique(asp_scots_p.val$OrthoGroup)) )

cat("Number of duplicated genes in Asp: ", sum(duplicated(asp_scots_p.val$Species1)), "\nNumber of duplicated genes in Scots: ",sum(duplicated(asp_scots_p.val$Species2)))

# AspWood filtering and sorting

asp_pval_copy <- asp_scots_p.val
asp_expr_fs <- asp_pval_copy %>% 
  left_join(asp_expr_data, join_by(Species1 == Genes)) %>% 
  select(-c("Species2", "Max.p.val", "OrthoGroup")) %>% 
  as_tibble() %>% 
  column_to_rownames(var = "Species1") %>% 
  select(contains("1."))
  
  
# ScotsWood filtering and sorting

scots_pval_copy <- asp_scots_p.val
scots_expr_fs <- scots_pval_copy %>% 
  left_join(scots_expr_data, join_by(Species2 == Genes)) %>% 
  select(-c("Species1", "Max.p.val", "OrthoGroup")) %>% 
  as_tibble() %>% 
  column_to_rownames(var = "Species2") %>% 
  select(contains("1."))

# Correlation amd heatmap
corr_asp_scots <- round(cor(asp_expr_fs, scots_expr_fs,method = "pearson"), 4)


heat_asp_scots <-  Heatmap(corr_asp_scots, column_title = "Asp-Scots",
    column_title_gp = gpar(fontsize = 15),
    row_names_gp = gpar(fontsize = 8),
    column_names_gp = gpar(fontsize = 8),
    width = ncol(corr_asp_scots)*unit(5, "mm"), 
    height = nrow(corr_asp_scots)*unit(5, "mm"),
    row_order = 1:ncol(asp_expr_fs),
    column_order = 1:ncol(scots_expr_fs),
    name = "corr") 

```



```{r ordinary plot}

heats <-  heat_asp_nor + heat_asp_scots
heats


```



```{r cowplot}

g1 <- grid.grabExpr(print(heat_asp_nor))
g2 <- grid.grabExpr(print(heat_asp_scots))
plot_grid(g1, g2)


```













