---
title: "GO enrichment"
author: "Ellen Dimmen Chapple"
date: '`r format(Sys.time(), "%d.%m.%Y")`'
output: 
  html_document:
    code_folding: hide
    toc: true
    number_sections: true
    toc_float: true
    theme: lumen
editor_options:
  chunk_output_type: console
---



```{r message=FALSE, warning=FALSE}

library(tidyverse)
library(ggplot2)
library(cowplot)

```


- NAC genes, vnd
- more functional info in arabidopsis
- use plantgenie


# Expression profiles

Flow: 
1) Find a GO term of interest - filter the go_list to only contain genes with this term OR look for previously found marker genes.

2) Filter the ortholog group file to only contain annotated genes for the specific GO term and that only occur in the gene set we are studying (ex. ultra-conserved genes). Fill in which gene set is to be used!

# Filtering by GO terms

First GO: GO:0009834 - plant-type secondary cell wall biogenesis
Second GO: GO:1903298 - negative regulation of hypoxia-induced intrinsic apoptotic signaling...
Third GO: GO:0045492 - xylan biosynthetic process

```{r message=FALSE, warning=FALSE}

# ---------------------------------------------------------------

load("summary_table_cons_0.05.RData")
summary_table <- summary_table_cons_0.05

species_list <- c("Lodge", "Asp", "Nor","Scots","Birch","Cher")

conserved_genes <- read_delim("Data/DATA/textFiles/cons_genes_aspen.txt", " ")

ortholog_group_file <- read.delim("Data/DATA/Orthogroups.100323.tsv", header = TRUE, sep = "\t")

co_expressed_genes <- read.delim("Data/DATA/textFiles/all_co-expressed_genes.txt", header = TRUE, sep = " ")

load("go_list.RData")

# # Only the terms that are enriched in the conserved and annotated genes.
# test <- go_list %>% 
#   filter(`Sequence Name` %in% conserved_genes$Asp)
#  
# length(unique(test$`Sequence Name`)) # Number of unique genes that are enriched in the conserved genes list (and are also annotated and significant).
# length(unique(test$`GO ID`)) # Why are there fewer GO ID's in test than in the summary table?
# length(unique(summary_table_cons_0.05$`GO id`))
# 
# check <- go_list %>% 
#   filter(`GO ID` %in% summary_table_cons_0.05$`GO id`)
# 
# length(unique(check$`GO ID`))

# ---------------------------------------------------------------

load("summary_table_cons_0.05.Rdata")
non_overlapping_GO <- summary_table %>% 
  filter(!(`GO id` %in% summary_table_cons_0.05$`GO id`))

paste0("Number of non-overlapping GO IDs: ", length(unique(non_overlapping_GO$`GO id`)) )

actual_GO_terms <- go_list %>% 
  filter(`GO ID` %in% summary_table$`GO id`) %>% 
  filter(`Sequence Name` %in% conserved_genes$Asp) %>% 
  left_join(summary_table, join_by(`GO ID` == `GO id`)) %>% 
  select(-c(4:7))

unique(actual_GO_terms$`GO term`)

filtered_sum_table <- summary_table %>% 
  filter(`GO id` %in% actual_GO_terms$`GO ID`) # Just to look at


go_input <- "GO:0042732"
go_term <- "D-xylose metabolic process"

# Filter the go-list to only contain asp genes associated with the GO id
goterm_to_genes <- actual_GO_terms %>% 
  filter(`GO ID` == go_input ) %>% 
  distinct(`Sequence Name`, .keep_all = T) # removes variants


genes_to_all_genes <- ortholog_group_file %>%  
  mutate(Pinsyl.SHORT.cp.pep = Pinsyl.SHORT.pep) %>% 
  rename(
    Asp = Poptre.SHORT.pep,
    Birch = Betpen.SHORT.pep,
    Nor = Picabi.SHORT.pep,
    Scots = Pinsyl.SHORT.pep,
    Cher = Prunavi.SHORT.pep,
    Lodge = Pinsyl.SHORT.cp.pep,
    OrthoGroup = Orthogroup) %>%   
  separate_rows(Asp, sep = ", ") %>% 
  mutate(Asp = gsub("Poptre_", "", Asp)) %>%
  mutate(Asp = gsub("\\.\\d\\.p\\d$", "", Asp)) %>% 
  select(OrthoGroup, all_of(species_list)) %>% 
  filter(Asp %in% goterm_to_genes$`Sequence Name`)%>% 
  separate_rows(Birch, sep = ", ") %>%
  separate_rows(Cher, sep = ", ") %>% 
  separate_rows(Nor, sep = ", ") %>% 
  separate_rows(Lodge, sep = ", ") %>% 
  separate_rows(Scots, sep = ", ") %>% 
  mutate(Birch = gsub("Betpen_", "", Birch)) %>%
  mutate(Cher = gsub("Prunavi_", "", Cher)) %>%
  mutate(Nor = gsub("\\.p\\d$", "", Nor)) %>%
  mutate(Cher = gsub("\\.p\\d$", "", Cher)) 



```



```{r}

file_list_2 <- c()
for (i in species_list){

  expr_name <- paste0("Data/DATA/transcriptomicsData/transcriptomicsForHeatmaps/",i,"Wood_transcriptomics_hm.txt")
  file_list_2 <- append(file_list_2, expr_name, after = length(file_list_2))

}


```



3) When a gene of interest has been identified, we filter the data set to only contain genes within this ortholog group. In advance the genes for Aspen with the same GO term have been plotted in order to evaluate if the gene is a good marker gene candidate. 

 **Comments**

From ver2:
Most genes share a similar curve where there is a peak around sample 12. These genes are not orthologs, but all share one GO. Before choosing one of these (and then going with this ortholog group), take a look at each gene in PlantGenie to see what other GOs the gene has.


According to PlantGenie, the following genes actually contain the GO:
 
 "Potra2n14c26547" **
 "Potra2n5c11175"  
 "Potra2n5c12319"  **
 "Potra2n7c15752" **

** Have good profiles 

```{r}

# genes_to_plot_filtered <- c( "Potra2n14c26547", "Potra2n5c12319" ) 
# # Other genes associated with the GO term:  "Potra2n14c26547" ,  "Potra2n5c12319" ,  "Potra2n7c15752"
# 
# aspGene_to_OG <- genes_to_all_genes %>%  
#   filter(Asp %in% genes_to_plot_filtered)

filtered_co_ex <- co_expressed_genes %>% 
  filter(MaxpVal < 0.05) %>% 
  select(GeneSpecies1, GeneSpecies2) 

# Create a vector to only contain co-expressed genes
gene_vector <- data.frame(genes = unlist(filtered_co_ex, use.names = FALSE))

gene_vector <- gene_vector %>% 
  distinct(genes, .keep_all = T)


# Filter the orthologs to only keep the genes that are significantly co-expressed

OG_to_genes <- genes_to_all_genes %>% 
  filter(Scots %in% gene_vector$genes) %>% 
  filter(Nor %in% gene_vector$genes) %>% 
  filter(Lodge %in% gene_vector$genes) %>% 
  filter(Asp %in% gene_vector$genes) %>% 
  filter(Cher %in% gene_vector$genes) %>% 
  filter(Birch %in% gene_vector$genes) %>% 
  group_by(OrthoGroup) 
  # %>% slice(1)

# Print out the unique orthologs for each species

for(i in 2:ncol(OG_to_genes)){
  col <- colnames(OG_to_genes[i])
  
  cat("Species: ", col, "\n")
  cat("Number of unique genes: ",nrow(unique(OG_to_genes[,i])), "\n")
  cat("Genes: ", "\n")
  print(unique(OG_to_genes[i]))
  cat("\n")
  cat("\n")
}




```




```{r fig.dim=c(15,15), message=FALSE, warning=FALSE}

expression_all_longer <- data.frame() 


for(i in 1:length(file_list_2)){

  

species <- sapply(strsplit(file_list_2[i], "/transcriptomicsForHeatmaps/"), "[",2)
species <- sapply(strsplit(species, "Wood"), "[", 1)    
  
filtering_vector <- OG_to_genes %>% 
  rename(Species = species)
  
  expression_vector <- read_delim(file_list_2[i], show_col_types = FALSE)  %>% 
    select(Genes, contains("1.")) %>% 
    filter(Genes %in% filtering_vector$Species)
  
  if(nrow(expression_vector > 0)){
    
  samples <- colnames(expression_vector)[-1]
  expression_vector_t <- t(expression_vector) 
  colnames(expression_vector_t) <- expression_vector_t[1,]
  

  expression_vector_t <- expression_vector_t[-1, ,drop = F] %>% 
    as.data.frame() %>%
    mutate(Samples = samples) 
  
  
  longer <- expression_vector_t %>% 
    pivot_longer(-Samples,names_to = "Genes", values_to = "Expression") %>% 
    mutate_at("Expression", as.numeric) %>% 
    mutate(Samples = gsub("^(.+?).", "", Samples)) 
  
  longer_end <- longer %>% 
    mutate(Species = rep(species, nrow(longer)))
  

  
  expression_all_longer <- rbind(expression_all_longer,longer_end)
  } else{
    print(paste0("Nothing to show for: ", species))
  }

  
  
}

expression_all_longer <- expression_all_longer %>% 
  group_by(Species)

plots <- list()

for(i in 1:length(unique(expression_all_longer$Species))){
  
  species <- unique(expression_all_longer$Species)[i]
  
  
  df <- expression_all_longer %>% 
    filter(Species == species)
  
  sub_title <- paste0(species)
  
  plots[[length(plots)+1]]  <- ggplot(data = df, aes(x = Samples, y = Expression, col = Genes ,group = Genes)) +
    geom_line(linewidth = 2) +
    ggtitle(sub_title) + ylab("Expression")
  
  
  
}

```

```{r fig.dim = c(15,15)}

plot_title <- ggdraw() + draw_label(paste0(go_input," - " ,go_term), fontface='bold')
plot_grid(plot_title, NULL,plotlist = plots, ncol = 2, nrow = 4) 

```


# Filtering by selected genes


Are these genes below amongst the ultra conserved genes? Amongst the less conserved genes?

"Potri.004G081300", "Potri.016G142800", "Potri.001G240900", "Potri.004G059600", "Potri.011G044500" - mapped to the triocarpa

("SUS6", "CDC2", "EXPA1", "CesA8", "BFN1") P. trichocarpa

The genes are not in the  conserved file.


```{r}

poptri_genes <- c("Potri.004G081300", "Potri.016G142800", "Potri.001G240900", "Potri.004G059600", "Potri.011G044500")

# OGs_from_jan23 <- c("OG0000548", "OG0002721", "OG0005186", "OG0009733" ,"OG0036093")

ortholog_file_filtered <- ortholog_group_file %>% 
  rename(Asp = Poptre.SHORT.pep,
         Tri = Poptri.SHORT.pep) %>% 
  separate_rows(Asp, sep = ", ") %>% 
  separate_rows(Tri, sep = ", ") %>% 
  mutate(Asp = gsub("Poptre_", "", Asp)) %>%
  mutate(Asp = gsub("\\.\\d\\.p\\d$", "", Asp)) %>% 
  mutate(Tri = gsub("Poptri_", "", Tri)) %>% 
  mutate(Tri = gsub("\\|.*", "", Tri)) %>% 
  mutate(Tri = gsub("\\.\\d$", "", Tri)) %>% 
  filter(Tri %in% poptri_genes) %>% 
  filter(Asp %in% conserved_genes$Asp)

length(unique(ortholog_file_filtered$Tri)) 

length(unique(ortholog_file_filtered$Asp)) 

  
```


```{r}

genes_to_all_genes <- ortholog_file_filtered %>%
  mutate(Pinsyl.SHORT.cp.pep = Pinsyl.SHORT.pep) %>%
  rename(
    Birch = Betpen.SHORT.pep,
    Nor = Picabi.SHORT.pep,
    Scots = Pinsyl.SHORT.pep,
    Cher = Prunavi.SHORT.pep,
    Lodge = Pinsyl.SHORT.cp.pep,
    OrthoGroup = Orthogroup) %>%
  select(OrthoGroup, Tri ,all_of(species_list)) 

```

```{r}

# removable_ogs <- c("OG0016249", "OG0000439")
#  filter(!(OrthoGroup %in% removable_ogs) ) %>% 

# Separate rows when there are fewer genes
OG_to_genes <- genes_to_all_genes %>% 

  separate_rows(Birch, sep = ", ") %>%
  separate_rows(Cher, sep = ", ") %>% 
  separate_rows(Nor, sep = ", ") %>% 
  separate_rows(Lodge, sep = ", ") %>% 
  separate_rows(Scots, sep = ", ") %>% 
 
  mutate(Birch = gsub("Betpen_", "", Birch)) %>%
  mutate(Cher = gsub("Prunavi_", "", Cher)) %>%
  mutate(Nor = gsub("\\.p\\d$", "", Nor)) %>%
  mutate(Cher = gsub("\\.p\\d$", "", Cher))

unique_genes_tri <- unique(OG_to_genes$Tri)

```


```{r fig.dim = c(15,15)}



for(i in 1:length(unique_genes_tri)){
  
  # We want a plot of 6 figures per gene
  expression_all_longer <- data.frame() 
  plots <- list()
  
  # For all species
  for(x in 1:length(file_list_2)){
    # x <- 1
    # i <- 7
    
    species <- sapply(strsplit(file_list_2[x], "transcriptomicsData/"), "[",2)
    species <- sapply(strsplit(species, "Wood"), "[", 1)  
    tri <- unique_genes_tri[i]
    
    # tri_gene <- OG_to_genes %>% 
    #   filter(Tri == tri) 
    # 
    # tri_gene_v <- c(tri_gene)[[1]][1]
    #   
    filtering_vector <- OG_to_genes %>%
      filter(Tri == tri) %>%
      rename(Species = species)
    
    
    expression_vector <- read_delim(file_list_2[x], show_col_types = FALSE)  %>% 
      select(Genes, contains("1.")) %>% 
      filter(Genes %in% filtering_vector$Species)
    
    if(nrow(expression_vector) > 0){
          samples <- colnames(expression_vector)[-1]
    expression_vector_t <- t(expression_vector) 
    colnames(expression_vector_t) <- expression_vector_t[1,]
    
    expression_vector_t <- expression_vector_t[-1, ,drop = F] %>% 
      as.data.frame() %>%
      mutate(Samples = samples) 
    
    
    longer <- expression_vector_t %>% 
      pivot_longer(-Samples,names_to = "Genes", values_to = "Expression") %>% 
      mutate_at("Expression", as.numeric) %>% 
      mutate(Samples = gsub("^(.+?).", "", Samples)) 
    
    longer_end <- longer %>% 
      mutate(Species = rep(species, nrow(longer)))
    
    expression_all_longer <- rbind(expression_all_longer,longer_end)
    } else{
      print(paste0("No ortholog of: ", tri, " for ", species))
    }
    
  }
  
  
  expression_all_longer <- expression_all_longer %>% 
    group_by(Species)
  
  
  
  for(q in 1:length(unique(expression_all_longer$Species))){
    
    species <- unique(expression_all_longer$Species)[q]
    
    
    df <- expression_all_longer %>% 
      filter(Species == species)
    
    sub_title <- paste0(species)
    
    plots[[length(plots)+1]]  <- 
      ggplot(data = df, aes(x = Samples, y = Expression, col = Genes ,group = Genes)) +
      geom_line(linewidth = 2) +
      ggtitle(sub_title) + ylab("Expression")
    
    
    
    
  }
  
  print(tri)
  
  plot_title <- ggdraw() + draw_label(tri, fontface='bold')
  print(plot_grid(plot_title, NULL, plotlist = plots, ncol = 2, nrow = 4))
  
}


```






