---
title: "Conserved I-graph"
author: "Ellen Dimmen Chapple"
date: "`r Sys.Date()`"
output: 
  html_document:
    code_folding: hide
    toc: true
    number_sections: true
    toc_float: true
    theme: lumen
editor_options:
  chunk_output_type: console
---


```{r}

library(tidyverse)
library(igraph)

```


```{r}
# Read in expressologs
expressologs <- read.delim("Data/DATA/textFiles/cons_genes_gymnosperms.txt", sep = " ") %>% 
  mutate(SpeciesPair = paste0(Species1, Species2))

# Set p-value threshold
p_thr <- 0.2

# Identify conservation orthogroups - all 15 species comparisons: utraconserved
speciesPairs <- tibble(SpeciesPair = expressologs %>% distinct(SpeciesPair) %>% pull(SpeciesPair))


```

Good too know: n() gives group size!

```{r}
expressologs <- left_join(expressologs, speciesPairs, by = "SpeciesPair")

# Not strictly necassasry, but practical for removing extra expressologs?
expressologs_groups <- expressologs %>% 
  group_by(OrthologGroup, SpeciesPair) %>% 
  arrange(MaxpVal) %>%
  filter(MaxpVal < p_thr) %>% 

  group_by(OrthologGroup) %>%
  mutate(Conserved = n()) %>%
  arrange(desc(Conserved)) %>% 
  ungroup()

ultraConservedOrthoGroups <- expressologs_groups %>% 
  group_by(OrthologGroup) %>% 
  distinct(SpeciesPair) %>% 
  distinct(OrthologGroup) %>% 
  pull(OrthologGroup)

cat("Orthogroups ultra-conserved amongst angiosperms:", length(ultraConservedOrthoGroups), "\n")

```



```{r}

# Identify ultraconserved orthogroups with one gene in each species: ultraultraconserved
# Make a network of pairwise expressologs and then look for cliques 

ultraultraConservedOrthoGroups <- c()
uuc_genes <- c()
for (g in ultraConservedOrthoGroups) {
  
  # g <- "OG0008027"
  
  expressologs_g <- expressologs_groups %>% 
    filter(OrthologGroup == g) 
  
  expressologs_g <- expressologs_g %>% 
    mutate(UGeneSpecies1 = paste0(expressologs_g$Species1, "-", expressologs_g$GeneSpecies1),
           UGeneSpecies2 = paste0(expressologs_g$Species2, "-", expressologs_g$GeneSpecies2))
  
  nodes <- data.frame(name = unique(c(expressologs_g$UGeneSpecies1, expressologs_g$UGeneSpecies2)))
  
  edges <- data.frame(from = expressologs_g$UGeneSpecies1, 
                      to = expressologs_g$UGeneSpecies2)
  
  net <- graph_from_data_frame(edges, directed = FALSE, vertices = nodes)
  
  
  # At around here we want to also include  'almost' cliques.
  
  clique <- largest_cliques(net)[[1]] # OBS: Just take one of them!
  clique <- attr(clique, "names")
  
  
  clique_species <- unique(sapply(str_split(clique, "-"), function(x) { x[1]}))
  clique_genes <-  str_split_fixed(clique, "-", n = 2)[,2]
  
  
  # Average number of edges per node:
  
  # degree_sum <- round(sum(degree(net))/length(names(degree(net))), 2)

  # cat("Ortholog group: ", g, "\n", "Average number of edges per node:  ", degree_sum)
  # cat("\n")
  # 
  # 
  
  # if(degree_sum >= 4.29){
  # 
  #   ultraultraConservedOrthoGroups <- c(ultraultraConservedOrthoGroups, g)
  # 
  # }
  
  if(length(clique_species) == 3) {
    
    ultraultraConservedOrthoGroups <- c(ultraultraConservedOrthoGroups, g)
    uuc_genes <- rbind(uuc_genes, data.frame(
      Orthogroup = rep(g, length(clique_genes)),
      Genes = clique_genes
      
      ))
      
      }}
      
  

cat("Number of ultra-ultra-conserved ortholog groups after identifying cliques based on inputs above: ",length(ultraultraConservedOrthoGroups))


```

Based on the out-print of average number of edges per node, it seems that the closest to a complete clique (5) is 4.29.

Using average of 5: 156 OGs 
Using average of 4.75: 156 OGs
Using average of 4.29: 274 OGs


Some plotting of expression profiles
Plot all genes in the orthogroup: should plot only those in the clique

```{r}

species_names <- c("NorWood", "ScotsWood", "LodgeWood")
species_names_short <- gsub("Wood", "", species_names)
tree_names <- c("S1", "T1", "L1")

orthoGroups <- ultraultraConservedOrthoGroups

expressologs %>% 
  filter(OrthologGroup %in% orthoGroups) -> expressologs_focus
genes <- unique(c(expressologs_focus$GeneSpecies1, expressologs_focus$GeneSpecies2))

expr <- data.frame()
for (species in species_names) {
  
  species_short <- gsub("Wood", "", species)
  tree <- tree_names[which(species == species_names)]
  
  species_expr_file <- paste0("Data/DATA/transcriptomicsData/transcriptomicsForHeatmaps/",species_short,"Wood_transcriptomics_hm.txt")
  species_expr <- read.delim(species_expr_file, sep = "\t", header = TRUE) %>% 
    filter(Genes %in% genes) %>% 
    gather (Sample.names, Expression, -1) %>%
    separate(Sample.names, into = c("Trees", "Samples"), sep = "\\.", remove = FALSE) %>%
    mutate_at("Samples", as.numeric) %>%
    filter(Trees == tree) %>% 
    mutate(Species = c(species_short))
  
  expr <- rbind(expr, species_expr)
  
}
```


```{r}

library(cowplot)

plots <- list()
for (i in 1:5) {
  
  expressologs %>% 
    filter(OrthologGroup %in% ultraultraConservedOrthoGroups[i]) -> expressologs_focusgroup
  orthoGroupGenes <- unique(c(expressologs_focusgroup$GeneSpecies1, expressologs_focusgroup$GeneSpecies2))
  
  plots[[length(plots)+1]] <- expr %>%
    filter(Genes %in% orthoGroupGenes) %>%
    ggplot(aes(x = Samples, y = Expression, col = Genes)) +
    geom_line(linewidth = 2) +
    ylab("Expression (VST)") +
    theme_classic() +
    theme(legend.position = "none") +
    facet_grid(cols = vars(Species)) +
    ggtitle(label = ultraultraConservedOrthoGroups[i])
  
}


plot_grid(plotlist = plots, ncol = 1)



```


```{r}

# Genes within a clique, uuc_genes
save(uuc_genes, file = "Data/DATA/uuc_orthologs_gymnosperms.RData")

```

```{r}

uuc_genes_asp <- uuc_genes %>% 
  filter(grepl("Potra", Genes))

save(uuc_genes_asp, file ="Data/DATA/uuc_orthologs_gymnosperms_asp.RData")

```


















