---
title: "GO enrichment"
author: "Ellen Dimmen Chapple"
date: '`r format(Sys.time(), "%d.%m.%Y")`'
output: 
  html_document:
    code_folding: hide
    toc: true
    number_sections: true
    toc_float: true
    theme: lumen
editor_options:
  chunk_output_type: console
---



```{r message=FALSE, warning=FALSE}

library(tidyverse)
library(ggplot2)
library(cowplot)

```


Flow: 
1) Load and filter the summary table to only contain GO IDs that are in the annotation file, then look through the output for any GO term that could be of interest.

2) Load the ortholog group file and the list of conserved and co-expressed genes. The list of conserved and co-expressed genes needs to be filtered by significant p-value.

3) Input the GO ID and GO Term and run code. The ortholog group file will now get filtered to only contain annotated genes for the specific GO ID. As we are using co-expressed genes we have already filtered for conserved ortholog groups.

# Filtering by GO terms

```{r message=FALSE, warning=FALSE}

# ORTHOLOG GROUPS
ortholog_group_file <- read.delim("Data/DATA/Orthogroups.100323.tsv", header = TRUE, sep = "\t")

# SUMMARY TABLE FROM GSEA
load("summary_table_cons_0.05.RData")
summary_table <- summary_table_cons_0.05 # Rename for conveniance 

print("Unique GO terms - all: ")
unique(summary_table$`GO term`)

# CONSERVED AND CO-EXPRESSED GENES - p-value cutoff at 0.05
# Create 2 separate columns of only aspen genes, then put them together to create one long column. Then remove the duplicated genes.

load("~/NMBU/M60-BIAS/EvoTree/githubEvoTree/Data/DATA/uuc_orthologs.RData")
conserved_genes <- uuc_genes

genes_s1_asp <-conserved_genes %>% 
  select(GeneSpecies1) %>% 
  filter(grepl("Potra",GeneSpecies1)) %>% 
  rename(gene = "GeneSpecies1")

genes_s2_asp <- conserved_genes %>% 
  select(GeneSpecies2) %>% 
  filter(grepl("Potra",GeneSpecies2)) %>% 
   rename(gene = "GeneSpecies2")

conserved_genes_only_aspen <- rbind(genes_s1_asp, genes_s2_asp)
conserved_genes_only_aspen <- conserved_genes_only_aspen %>% 
  distinct(gene, .keep_all = T)

# ANNOTATION FILE
load("go_list.RData")


# Filter away higher level GO terms that are not in annotation file
actual_GO_terms <- go_list %>% 
  filter(`GO ID` %in% summary_table$`GO id`) %>% 
  filter(`Sequence Name` %in% conserved_genes_only_aspen$gene) %>% 
  left_join(summary_table, join_by(`GO ID` == `GO id`)) %>% 
  select(-c(4:7))

save(actual_GO_terms, file = "Data/DATA/actual_GO_terms.R")

# Check out the unique GO terms
print("Unique GO terms - annotated: ")
unique(actual_GO_terms$`GO term`)



```


Choose a GO term and fill inn chunk below.

```{r}
go_input <- "GO:0016192"
go_term <- "vesicle-mediated transport"

```


Run code.



```{r}
species_list <- c("Lodge", "Asp", "Nor","Scots","Birch","Cher")

file_list_2 <- c()
for (i in species_list){

  expr_name <- paste0("Data/DATA/transcriptomicsData/transcriptomicsForHeatmaps/",i,"Wood_transcriptomics_hm.txt")
  file_list_2 <- append(file_list_2, expr_name, after = length(file_list_2))

}


```



```{r}

# Filter the go-list to only contain asp genes associated with the GO id
goterm_to_genes <- actual_GO_terms %>% 
  filter(`GO ID` == go_input ) %>% 
  distinct(`Sequence Name`, .keep_all = T) # removes variants

# Then filter the ortholog group file to only contain the orthologs of the aspen genes associated with the GO ID we have as input.

genes_to_all_genes <- ortholog_group_file %>%  
  mutate(Pinsyl.SHORT.cp.pep = Pinsyl.SHORT.pep) %>% 
  rename(
    Asp = Poptre.SHORT.pep,
    Birch = Betpen.SHORT.pep,
    Nor = Picabi.SHORT.pep,
    Scots = Pinsyl.SHORT.pep,
    Cher = Prunavi.SHORT.pep,
    Lodge = Pinsyl.SHORT.cp.pep,
    OrthoGroup = Orthogroup) %>%   
  separate_rows(Asp, sep = ", ") %>% 
  mutate(Asp = gsub("Poptre_", "", Asp)) %>%
  mutate(Asp = gsub("\\.\\d\\.p\\d$", "", Asp)) %>% 
  select(OrthoGroup, all_of(species_list)) %>% 
  filter(Asp %in% goterm_to_genes$`Sequence Name`)%>% 
  separate_rows(Birch, sep = ", ") %>%
  separate_rows(Cher, sep = ", ") %>% 
  separate_rows(Nor, sep = ", ") %>% 
  separate_rows(Lodge, sep = ", ") %>% 
  separate_rows(Scots, sep = ", ") %>% 
  mutate(Birch = gsub("Betpen_", "", Birch)) %>%
  mutate(Cher = gsub("Prunavi_", "", Cher)) %>%
  mutate(Nor = gsub("\\.p\\d$", "", Nor)) %>%
  mutate(Cher = gsub("\\.p\\d$", "", Cher)) 

# Next step is to filter the orthologs to only contain genes that are co-expressed.

all_co_ex_genes <- conserved_genes %>% 
  select(GeneSpecies1, GeneSpecies2)

gene_vector <- data.frame(genes = unlist(all_co_ex_genes, use.names = FALSE))

gene_vector <- gene_vector %>% 
  distinct(genes, .keep_all = T)

OG_to_genes <- genes_to_all_genes %>% 
  filter(Scots %in% gene_vector$genes) %>% 
  filter(Nor %in% gene_vector$genes) %>% 
  filter(Lodge %in% gene_vector$genes) %>% 
  filter(Asp %in% gene_vector$genes) %>% 
  filter(Cher %in% gene_vector$genes) %>% 
  filter(Birch %in% gene_vector$genes) %>% 
  group_by(OrthoGroup) 

```


By filtering the orthologs of the aspen genes that we isolated for the particular GO ID of interest with the gene_vector, we are selecting the genes (for all species) that actually are co-expressed and significant.


```{r}

for(i in 2:ncol(genes_to_all_genes)){
  col <- colnames(genes_to_all_genes[i])
  
  cat("Species: ", col, "\n")
  cat("Number of unique genes: ",nrow(unique(genes_to_all_genes[,i])), "\n")
  cat("Genes: ", "\n")
  print(unique(genes_to_all_genes[i]))
  cat("\n")
  cat("\n")

}


# Prints out the orthologs for each species

for(i in 2:ncol(OG_to_genes)){
  col <- colnames(OG_to_genes[i])
  
  cat("Species: ", col, "\n")
  cat("Number of unique genes: ",nrow(unique(OG_to_genes[,i])), "\n")
  cat("Genes: ", "\n")
  print(unique(OG_to_genes[i]))
  cat("\n")
  cat("\n")
}




```




```{r fig.dim=c(15,15), message=FALSE, warning=FALSE}

expression_all_longer <- data.frame() 


for(i in 1:length(file_list_2)){

  

species <- sapply(strsplit(file_list_2[i], "/transcriptomicsForHeatmaps/"), "[",2)
species <- sapply(strsplit(species, "Wood"), "[", 1)    
  
filtering_vector <- OG_to_genes %>% 
  rename(Species = species)
  
  expression_vector <- read_delim(file_list_2[i], show_col_types = FALSE)  %>%     select(Genes, contains("1.")) %>% 
    filter(Genes %in% filtering_vector$Species)
  
  if(nrow(expression_vector > 0)){
    
  samples <- colnames(expression_vector)[-1]
  expression_vector_t <- t(expression_vector) 
  colnames(expression_vector_t) <- expression_vector_t[1,]
  

  expression_vector_t <- expression_vector_t[-1, ,drop = F] %>% 
    as.data.frame() %>%
    mutate(Samples = samples) 
  
  
  longer <- expression_vector_t %>% 
    pivot_longer(-Samples,names_to = "Genes", values_to = "Expression") %>% 
    mutate_at("Expression", as.numeric) %>% 
    mutate(Samples = gsub("^(.+?).", "", Samples)) 
  
  longer_end <- longer %>% 
    mutate(Species = rep(species, nrow(longer)))
  

  
  expression_all_longer <- rbind(expression_all_longer,longer_end)
  } else{
    print(paste0("Nothing to show for: ", species))
  }

  
  
}

expression_all_longer <- expression_all_longer %>% 
  group_by(Species)

plots <- list()

for(i in 1:length(unique(expression_all_longer$Species))){
  
  species <- unique(expression_all_longer$Species)[i]
  
  
  df <- expression_all_longer %>% 
    filter(Species == species)
  
  sub_title <- paste0(species)
  
  plots[[length(plots)+1]]  <- ggplot(data = df, aes(x = Samples, y = Expression, col = Genes ,group = Genes)) +
    geom_line(linewidth = 2) +
    ggtitle(sub_title) + ylab("Expression")
  
  
  
}

```

```{r fig.dim = c(15,15)}

plot_title <- ggdraw() + draw_label(paste0(go_input," - " ,go_term), fontface='bold')
plot_grid(plot_title, NULL,plotlist = plots, ncol = 2, nrow = 4) 

```











