---
title: "GO enrichment"
author: "Ellen Dimmen Chapple"
date: '`r format(Sys.time(), "%d.%m.%Y")`'
output: 
  html_document:
    code_folding: hide
    toc: true
    number_sections: true
    toc_float: true
    theme: lumen
editor_options:
  chunk_output_type: console
---



```{r message=FALSE, warning=FALSE}

library(tidyverse)

library(GSEABase)
library(GOstats)
library(gdata)

```

# Gene Set Enrichment Analysis

- N, all expressed genes
- n, all expressed and annotated genes
- k, set of conserved genes
- x, genes that are conserved and annotated

## For a list of conserved genesx <- 

```{r eval=FALSE, message=FALSE, warning=FALSE, include=FALSE}

# # EXPRESSED GENES
# expressed_genes <- read_delim("Data/DATA/transcriptomicsData/AspWood_transcriptomics.txt")
# 
# # EXPRESSED GENES WITH ANNOTATION
# go_list_file <- read_delim("Data/DATA/Potra22_blast2go_GO.txt",'\t')
# 
# # GO
# go_list <- go_list_file %>% 
#     separate(`Sequence Name`,into = c("Sequence Name","Sequence Variant"), sep = "\\.") %>%
#     separate(`Annotation GO ID-Annotation GO Term`,into = c("GO ID","GO Term"), sep = "-", extra = "drop") %>%
#     filter(`Sequence Name` %in% expressed_genes$Genes)
# 
# go_data_frame <- data.frame(go_id=go_list$`GO ID`, evidence=rep("ND", nrow(go_list)), 
#                           gene_id=go_list$`Sequence Name`)
# go_frame_object=GOFrame(go_data_frame,organism="Aspen")
# go_all_frame_object=GOAllFrame(go_frame_object)
# 
# cat("Number of expressed genes: ", length(unique(expressed_genes$Genes)))
# cat("Number of expressed genes with at least 1 GO-term: ", length(unique(go_list$`Sequence Name`)))
# cat("Number of GO-terms: ", length(unique(go_list$`GO Term`)))
# cat("Number of unique genes in main annotation file: ", length(unique(go_list_file$`Sequence Name`)))
# 
# # PARAMETERS
# gene_set_collection <- GeneSetCollection(go_all_frame_object, setType = GOCollection())
# universe <- expressed_genes$Genes
# 
# # FILES WITH GENES CONSERVED BETWEEN CLADES AND SPECIFIC FOR CLADES - ASPEN GENES
# list <- c("Data/DATA/textFiles/cons_genes_aspen.txt", "Data/DATA/textFiles/gymnosperms_genes_aspen.txt", "Data/DATA/textFiles/angiosperms_genes_aspen.txt", "Data/DATA/textFiles/partial_3_genes_aspen.txt" )
# 
# 
# pval_cutoff <- 0.05
# for (i in 1:length(list)){
#   
# 
#  gene_set <- read_delim(list[i], " ")
# 
#   
#   gene_ID <- gene_set$Genes
# 
#   group_name <- sapply(strsplit(list[i], "textFiles/"), "[", 2)
#   group_name <- sapply(strsplit(group_name, "_genes"), "[", 1)
#   # 
#   params <- GSEAGOHyperGParams(name="GSEA",
#                                geneSetCollection = gene_set_collection,
#                                geneIds = gene_ID,
#                                universeGeneIds = universe, 
#                                ontology = "BP",
#                                pvalueCutoff = pval_cutoff,
#                                conditional = FALSE,
#                                testDirection = "over")
#   
#   hyper_g_test <- hyperGTest(params)
#   summary_table <- summary(hyper_g_test)
#   summary_table <- summary_table[summary_table$Count>1,]
#   summary_table <- summary_table[,c(1,2,5,6,7)]
#   colnames(summary_table) <- c("GO id","P-value","x","n","GO term")
#   
#   rownames(summary_table) <- 1:nrow(summary_table)
#   
#   summary_table$`P-value`<- format(summary_table$`P-value`, digits=3, scientific=TRUE)
#  
# 
#    
#   # Rename the summary table!
#   
#     new_name <- as.character(paste0("summary_table_", group_name, "_", pval_cutoff))
#     mv("summary_table", new_name) 
#     
#   
# }


```

## For a single file

```{r}
# EXPRESSED GENES
expressed_genes <- read_delim("Data/DATA/transcriptomicsData/AspWood_transcriptomics.txt")

# Conserved genes fill inn ---------------------------------------

load("Data/DATA/uuc_orthologs_asp.RData") # (uuc_orthologs_angiosperms_asp.RData OR uuc_orthologs_asp.RData)
# OR 
partial_3_genes_aspen <- read_delim("Data/DATA/textFiles/partial_3_genes_aspen.txt", " ") %>% 
  rename(Genes = "Asp")


gene_set <-  uuc_genes_asp
# Alter this depending on which gene set is used. Either: partial_3_genes_aspen or uuc_genes_asp


angiosperms <- "no"

pval_cutoff <- 0.05


#--------------------------------------------------------------------
# EXPRESSED GENES WITH ANNOTATION
go_list_file <- read_delim("Data/DATA/Potra22_blast2go_GO.txt",'\t')

# GO
go_list <- go_list_file %>% 
  separate(`Sequence Name`,into = c("Sequence Name","Sequence Variant"), sep = "\\.") %>%
  separate(`Annotation GO ID-Annotation GO Term`,into = c("GO ID","GO Term"), sep = "-", extra = "drop") %>%
  filter(`Sequence Name` %in% expressed_genes$Genes)

go_data_frame <- data.frame(go_id=go_list$`GO ID`, evidence=rep("ND", nrow(go_list)), 
                            gene_id=go_list$`Sequence Name`)

go_frame_object = GOFrame(go_data_frame,organism="Aspen")
go_all_frame_object = GOAllFrame(go_frame_object)

cat("Number of expressed genes: ", length(unique(expressed_genes$Genes)))
cat("Number of expressed genes with at least 1 GO-term: ", length(unique(go_list$`Sequence Name`)))
cat("Number of GO-terms: ", length(unique(go_list$`GO Term`)))
cat("Number of unique genes in main annotation file: ", length(unique(go_list_file$`Sequence Name`)))

# PARAMETERS
gene_set_collection <- GeneSetCollection(go_all_frame_object, setType = GOCollection())
universe <- expressed_genes$Genes

# FILE WITH ASPEN GENES FROM THE ULTRA-ULTRA-CONSERVED ORTHOLOG GROUPS (from IGRAPH)
----------------------------- GSEA ------------------------------------------------------------------------------

 # Saving asp genes for Plantgenie 

# vect <- gene_set$Genes
# vectStr=paste(as.character(vect), sep="' '", collapse=",")
# sink("gene_set_asp_uuc.txt")
# cat(vectStr)
# cat("\n")
# sink()
  
gene_ID <- gene_set$Genes
group_name <- "uuc_genes"

params <- GSEAGOHyperGParams(name="GSEA",
                             geneSetCollection = gene_set_collection,
                             geneIds = gene_ID,
                             universeGeneIds = universe, 
                             ontology = "BP",
                             pvalueCutoff = pval_cutoff,
                             conditional = FALSE,
                             testDirection = "over")

hyper_g_test <- hyperGTest(params)
summary_table_uuc <- summary(hyper_g_test)
summary_table_uuc <- summary_table_uuc[summary_table_uuc$Count>1,]
summary_table_uuc <- summary_table_uuc[,c(1,2,5,6,7)]
colnames(summary_table_uuc) <- c("GO id","P-value","x","n","GO term")

rownames(summary_table_uuc) <- 1:nrow(summary_table_uuc)

summary_table_uuc$`P-value`<- format(summary_table_uuc$`P-value`, digits=3, scientific=TRUE)

if(angiosperms == "yes"){
  
  summary_table_uuc_angio <- summary_table_uuc
}
  


```




```{r eval=FALSE, include=FALSE}

# save(go_list, file = "go_list.RData")
save(summary_table_uuc, file = "summary_table_uuc.RData")
save(summary_table_uuc_angio, file = "summary_table_uuc_angio.RData")



```
 
 
 

```{r eval=FALSE, include=FALSE}


wood_formation_terms <- c("wall","NAC" ,"phloem","lignin", "biosyn" ,"lignification","apoptosis", "apoptotic" ,"expansion", "growth", "camb","death", "cellulose", "xylem", "phloem", "ribosom", "transcrip", "transl")

summary_table_partial_3_keywords <- data.frame()
 
for (i in 1:length(wood_formation_terms)){
  
  word <- wood_formation_terms[i]
  df <- summary_table_partial_3_0.05 %>% 
    filter(grepl(word, `GO term`))
  
  if(nrow(df)>=1){
    df <- df %>% 
    mutate(key_word = word)
    
  summary_table_partial_3_keywords <- rbind(summary_table_partial_3_keywords, data.frame(df)) %>% 
    arrange(`key_word`)
  
  }
  
}


view(summary_table_partial_3_keywords)
cat("Number of unique id's: ", length(unique(summary_table_partial_3_keywords$GO.id)), " vs. nrows: ", nrow(summary_table_partial_3_keywords))



summary_table_ultra_cons_keywords <- data.frame()
 
for (i in 1:length(wood_formation_terms)){
  
  word <- wood_formation_terms[i]
  df <- summary_table_cons_0.05 %>% 
    filter(grepl(word, `GO term`))
  
  if(nrow(df)>=1){
    df <- df %>% 
    mutate(key_word = word)
    
  summary_table_ultra_cons_keywords <- rbind(summary_table_ultra_cons_keywords, data.frame(df)) %>% 
    arrange(`key_word`)
  
  }
  
}


view(summary_table_ultra_cons_keywords)
cat("Number of unique id's: ", length(unique(summary_table_ultra_cons_keywords$GO.id)), " vs. nrows: ", nrow(summary_table_ultra_cons_keywords))

```



```{r}
# Print GO IDs for REVIGO - use the R script created for plotting treemap.

df <- summary_table_uuc[,1]

for (i in df){
  cat(i, "\n")
}


```




